bpf/flow_monitor.h

```diff
@@ -0,0 +1,128 @@
+        aggregate_flow->bytes += skb->len;
+        aggregate_flow->end_mono_time_ts = current_time;
+        aggregate_flow->flags |= pkt.flags;
+
+        // Does not matter the gate. Will be zero if not enabled.
+        if (pkt.rtt > 0) {
+            /* Since RTT is calculated for few packets we need to check if it is non zero value then only we update
+             * the flow. If we remove this check a packet which fails to calculate RTT will override the previous valid
+             * RTT with 0.
+             */
+            aggregate_flow->flow_rtt = pkt.rtt;
+        }
+
+        long ret = bpf_map_update_elem(&aggregated_flows, &id, aggregate_flow, BPF_ANY);
+        if (trace_messages && ret != 0) {
+            // usually error -16 (-EBUSY) is printed here.
+            // In this case, the flow is dropped, as submitting it to the ringbuffer would cause
+            // a duplicated UNION of flows (two different flows with partial aggregation of the same packets),
+            // which can't be deduplicated.
+            // other possible values https://chromium.googlesource.com/chromiumos/docs/+/master/constants/errnos.md
+            bpf_printk("error updating flow %d\n", ret);
+        }
+    } else {
+        // Key does not exist in the map, and will need to create a new entry.
+        flow_metrics new_flow = {
+            .packets = 1,
+            .bytes = skb->len,
+            .start_mono_time_ts = current_time,
+            .end_mono_time_ts = current_time,
+            .flags = pkt.flags,
+            .flow_rtt = pkt.rtt
+        };
+
+        // even if we know that the entry is new, another CPU might be concurrently inserting a flow
+        // so we need to specify BPF_ANY
+        long ret = bpf_map_update_elem(&aggregated_flows, &id, &new_flow, BPF_ANY);
+        if (ret != 0) {
+            // usually error -16 (-EBUSY) or -7 (E2BIG) is printed here.
+            // In this case, we send the single-packet flow via ringbuffer as in the worst case we can have
+            // a repeated INTERSECTION of flows (different flows aggregating different packets),
+            // which can be re-aggregated at userpace.
+            // other possible values https://chromium.googlesource.com/chromiumos/docs/+/master/constants/errnos.md
+            if (trace_messages) {
+                bpf_printk("error adding flow %d\n", ret);
+            }
+
+            new_flow.errno = -ret;
+            flow_record *record = (flow_record *)bpf_ringbuf_reserve(&direct_flows, sizeof(flow_record), 0);
+            if (!record) {
+                if (trace_messages) {
+                    bpf_printk("couldn't reserve space in the ringbuf. Dropping flow");
+                }
+                return TC_ACT_OK;
+            }
+            record->id = id;
+            record->metrics = new_flow;
+            bpf_ringbuf_submit(record, 0);
+        }
+    }
+    return TC_ACT_OK;
+}
+
+#endif /* __FLOW_MONITOR_H__ */
```

Comment:
nit: pls add new line here

Reply:
Done. Understood the style, will follow it.